### PR TITLE
feat: improvements into app structure

### LINKED ISSUE #1: App Structure: Minor improvements needed
Hey! I truly liked the concept of the project and sure it will be helpful on my studies.

When I was testing it I found out a few things to improve on the app structure, to give more readability and use some types:

- Recover the Progress in the ListVideos page and not only at ContinueVideos;
  - Since everytime that I join a video that I started to watch, it goes back to zero and it's kinda annoying. 
- Add validations if there's no video on ContinueVideos list. Now it's crashing if there's no video. 
- Missing a core type for `Video` and `WatchHistory`;
- Segregate Back-end from Front-end (I think that it will improve it a lot);
- Resize cards to a minimum width since there's huge titles to cover;
- Replace all statements to PreparedStataments;
- Give some cool spacing for code readability;
- Add a progress bar into ListVideos and ContinueVideos page.

Probably there's many more things that I missed but the app itself works perfectly!

I started to work on these same points btw, I'll open a draft for now and we can check it together!

### PR DESCRIPTION
**Pull request by danielhe4rt**
Resolves #1 

- [x] Recover the Progress in the ListVideos page and not only at ContinueVideos;
  - Since everytime that I join a video that I started to watch, it goes back to zero and it's kinda annoying. 
- [x] Add validations if there's no video on ContinueVideos list. Now it's crashing if there's no video. 
- [x] Missing a core type for `Video` and `WatchHistory`;
- [x] Segregate Back-end from Front-end (I think that it will improve it a lot);
- [x] Resize cards to a minimum width since there's huge titles to cover;
- [x] Replace all statements to PreparedStataments;
- [x] Give some cool spacing for code readability;
- [x] Add a progress bar into ListVideos and ContinueVideos page.

### CODE REVIEW

File: src/pages/api/list-continue.ts [hunk 1]
```diff
@@ -0,0 +1,50 @@
...
+            duration: row.duration,
+            url: row.url,
+            progress: watchedVideoProgress.get(row.id)
+        };
+
+        videos.push(video)
+    }
+
+    return res.status(200).json(videos);
+}
```
Comment (Canhassi12) [inline src/pages/api/list-continue.ts:33]:
```suggestion
}

```

File: src/components/VideoCard.tsx [hunk 2]
```diff
@@ -1,18 +1,22 @@
...
 import CardContent from '@mui/material/CardContent';
-import Button from '@mui/material/Button';
 import Typography from '@mui/material/Typography';
+import Image from 'next/image';
+import { Video } from 'src/types';
+
+export default function VideoCard({video}: {video: Video}) {
+  const watchUrl = `/watch/${video.video_id}?progress=${video.progress}`
+  const progressPercentage = Math.floor((video.progress / video.duration) * 100);
+  console.log(video);
```
Comment (JVictorV) [inline src/components/VideoCard.tsx]:
```suggestion
```

File: src/types/index.d.ts [hunk 3]
```diff
@@ -0,0 +1,18 @@
...
+    url: string,
+    duration: internal,
+    progress: ?number
+}
+
+export interface WatchHistory {
+    user_id: userId,
+    video_id: videoId,
+    progress: progress,
+    watched_at: watchedAt
```
Comment (JVictorV) [inline src/types/index.d.ts]: Use proper types here

File: src/types/index.d.ts [hunk 4]
```diff
@@ -0,0 +1,18 @@
+import internal from "stream"
+
+export interface Video {
+    video_id: string,
+    thumbnail: string,
+    title: string,
+    content_type: string,
+    url: string,
+    duration: internal,
```
Comment (JVictorV) [inline src/types/index.d.ts]: Internal doesnt seem right here, that should probably be a number?
```suggestion
    duration: number,
```

File: src/types/index.d.ts [hunk 5]
```diff
@@ -0,0 +1,18 @@
+import internal from "stream"
```
Comment (JVictorV) [inline src/types/index.d.ts]:
```suggestion
```

File: src/pages/watch/[videoId].tsx [hunk 6]
```diff
@@ -104,10 +96,16 @@ export default function Watch({ video }: Video) {
...
-                        <h1>{video.title}</h1>
-                        <video controls ref={videoRef} className="video-js" />
+                        <div style={{
+                            display: 'flex',
+                            alignItems: 'center',
+                            flexDirection: 'column'
+                        }}>
+                            <h1>{video.title}</h1>
+                            <video controls ref={videoRef} className="video-js" height={'500'} />
+                        </div>
```
Comment (JVictorV) [inline src/pages/watch/[videoId].tsx]: You could use the proper [Box](https://mui.com/material-ui/react-box/) and [Typography](https://mui.com/material-ui/react-typography/) components here

File: src/pages/watch/[videoId].tsx [hunk 7]
```diff
@@ -1,49 +1,41 @@
...
-                url: results.rows[0].url,
-                progress: progress
-            }
-        }
+    const video: Video = {
+        video_id: results.rows[0].id,
+        thumbnail: results.rows[0].thumbnail,
+        title: results.rows[0].title,
+        content_type: results.rows[0].content_type,
+        url: results.rows[0].url,
```
Comment (JVictorV) [inline src/pages/watch/[videoId].tsx]: ```results.rows[0]``` here could be undefined, doing some validation would be nice

File: src/pages/index.tsx [hunk 8]
```diff
@@ -1,39 +1,33 @@
...
   return {
     props: {
-      videos: data
+      videos: videosResponse
     }
   }
 }
 
-export default function AllVideos({ videos }) {
+export default function AllVideos({ videos }: {videos: Video[]}) {
```
Comment (JVictorV) [inline src/pages/index.tsx]:
```suggestion
interface AllVideosProps {
  videos: Video[]
}
export default function AllVideos({ videos }: AllVideosProps) {
```

File: src/pages/continue.tsx [hunk 9]
```diff
@@ -1,56 +1,45 @@
...
   return {
-    props: {
-      videos: data
-    }
+    props: { videos: videosResponse }
   }
 }
 
-export default function ContinueWatching({ videos }) {
+function noVideoAvailable() {
```
Comment (JVictorV) [inline src/pages/continue.tsx]: React components should always use pascal case
```suggestion
function NoVideoAvailable() {
```

File: src/pages/continue.tsx [hunk 10]
```diff
@@ -66,16 +55,15 @@ export default function ContinueWatching({ videos }) {
         >
           <Toolbar />
           <Container maxWidth="lg" sx={{ mt: 4, mb: 4 }}>
+            <h1>List of videos which you started</h1>
+
+            {videos.length === 0 ? noVideoAvailable() : ''}
```
Comment (JVictorV) [inline src/pages/continue.tsx]:
```suggestion
            {videos.length === 0 ? <NoVideoAvailable /> : null}
```

File: src/pages/api/list-videos.ts [hunk 11]
```diff
@@ -0,0 +1,33 @@
...
+
+        videoCollection.push(video);
+    }
+    
+    return res.status(200).json(videoCollection);
+}
+
+function getProgress(progress) {
+    return progress.rows.length > 0 ? progress.rows[0].progress : 0;
+}
```
Comment (JVictorV) [inline src/pages/api/list-videos.ts]:
```suggestion
function getProgress(progress: types.ResultSet) {
    return progress.first().get('progress') ?? 0;
}
```

File: src/pages/api/list-continue.ts [hunk 12]
```diff
@@ -0,0 +1,50 @@
...
+
+    // fetch video IDs
+    const watchedVideos = await cluster.execute("SELECT video_id, progress FROM watch_history WHERE user_id = ? LIMIT 9;", [userId])
+
+    if (watchedVideos.rows.length === 0) {
+        return res.status(200).json([]);
+    }
+
+
+    let watchedVideoProgress = new Map()
```
Comment (JVictorV) [inline src/pages/api/list-continue.ts]:
```suggestion
    let watchedVideoProgress = new Map<string, number>()
```

**Review by JVictorV — CHANGES_REQUESTED** [review]

File: src/types/index.d.ts (hunk 3 above)
Comment (danielhe4rt) [reply src/types/index.d.ts]: What do you mean with proper types?

File: src/types/index.d.ts (hunk 4 above)
Comment (danielhe4rt) [reply src/types/index.d.ts]: Fixed.

File: src/db/scylladb.ts [hunk 13]
```diff
@@ -1,10 +1,29 @@
...
+      thumbnail: video.thumbnail,
+      title: video.title,
+      content_type: video.content_type,
+      duration: video.duration,
+      url: video.url,
+      progress: progress
+  };
+
+  return parsedVideo;
+}
```
Comment (Canhassi12) [inline src/db/scylladb.ts:29]:
```suggestion
}

```

File: src/pages/api/list-videos.ts [hunk 14]
```diff
@@ -0,0 +1,28 @@
...
+        
+        return parseVideo(rawVideo, getProgress(progress))
+    });
+    
+    return res.status(200).json(await Promise.all(videoCollection));
+}
+
+function getProgress(progress: types.ResultSet) {
+    return progress.first().get('progress') ?? 0;
+}
```
Comment (Canhassi12) [inline src/pages/api/list-videos.ts:34]:
```suggestion
}

```

File: src/pages/api/videos/[videoId].ts [hunk 15]
```diff
@@ -0,0 +1,21 @@
...
+    const result = await cluster.execute(`SELECT * FROM streaming.video WHERE id = ?`, [req.query.videoId]);
+    const fetchedVideo = result.rows[0];
+    if (!fetchedVideo) {
+        return res.status(404);
+    }
+
+    const video = parseVideo(fetchedVideo, 0)
+
+    return res.status(200).json(video)
+}
```
Comment (Canhassi12) [inline src/pages/api/videos/[videoId].ts:21]:
```suggestion
}

```

File: migrate/sample_data.csv [hunk 16]
```diff
@@ -1,9 +1,9 @@
...
+scylladb-no-code-changes,2023-09-06 17:21:24,video/youtube,https://img.youtube.com/vi/Mr7VzZNyGW4/mqdefault.jpg,SAS Shifts Apps to ScyllaDB with No Code Changes,https://www.youtube.com/watch?v=Mr7VzZNyGW4,123
+scylladb-cloud-terraform-2,2023-09-06 16:18:23,video/youtube,https://img.youtube.com/vi/sA8Rqpggr30/mqdefault.jpg,ScyllaDB Cloud Terraform provider #2: Create new cluster,https://www.youtube.com/watch?v=sA8Rqpggr30,507
+future-of-nosql,2023-09-05 17:28:02,video/youtube,https://img.youtube.com/vi/GMK3Hpe2VDM/mqdefault.jpg,Expert Panel The Future of NoSQL Databases,https://www.youtube.com/watch?v=GMK3Hpe2VDM,3382
+how-discord-migrated-trillions-of-messages,2023-09-06 17:23:24,video/youtube,https://img.youtube.com/vi/S2xmFOAUhsk/mqdefault.jpg,How Discord Migrated Trillions of Messages from Cassandra to ScyllaDB,https://www.youtube.com/watch?v=S2xmFOAUhsk,1699
+whats-new-scylladb-5,2023-09-06 16:19:23,video/youtube,https://img.youtube.com/vi/gZE73loJLKY/mqdefault.jpg,What’s New in ScyllaDB Open Source 5.0,https://www.youtube.com/watch?v=gZE73loJLKY,3242
```
Comment (Canhassi12) [inline migrate/sample_data.csv:9]: you write this data manually?

Comment (danielhe4rt) [reply migrate/sample_data.csv:9]: This is just a sample data. No need to be fetching from an YouTube integration.

At this specific point, I'm using this slug to prettify how the video appears in the browser URL.

Comment (Canhassi12) [reply migrate/sample_data.csv:9]: I got it.

**Comment by zseta** [comment]
Thanks @DanielHe4rt for the PR! I can't run the app with your modifications, I get multiple errors, eg:
```
- error src/pages/api/list-videos.ts (27:28) @ get
- error Error [TypeError]: Cannot read properties of null (reading 'get')
    at getProgress (webpack-internal:///(api)/./src/pages/api/list-videos.ts:24:28)
    at eval (webpack-internal:///(api)/./src/pages/api/list-videos.ts:19:86)
    at process.processTicksAndRejections (node:internal/process/task_queues:95:5)
    at async Promise.all (index 7)
    at async listVideos (webpack-internal:///(api)/./src/pages/api/list-videos.ts:21:33) {
  digest: undefined
}
  25 | 
  26 | function getProgress(progress: types.ResultSet) {
> 27 |     return progress.first().get('progress') ?? 0;
     |                            ^
  28 | }
- wait compiling /_error (client and server)...
- event compiled client and server successfully in 151 ms (1330 modules)
- error Error [SyntaxError]: Unexpected token '<', "<!DOCTYPE "... is not valid JSON
    at JSON.parse (<anonymous>)
    at parseJSONFromBytes (node:internal/deps/undici/undici:6662:19)
    at successSteps (node:internal/deps/undici/undici:6636:27)
    at node:internal/deps/undici/undici:1236:60
    at node:internal/process/task_queues:140:7
    at AsyncResource.runInAsyncScope (node:async_hooks:206:9)
    at AsyncResource.runMicrotask (node:internal/process/task_queues:137:8)
    at process.processTicksAndRejections (node:internal/process/task_queues:95:5) {
  digest: undefined
```
Could you test again maybe with a fresh install?

**Comment by danielhe4rt** [comment]
@zseta Fixed!

**Comment by zseta** [comment]
I still get this error:
```
✔ Next.js was successfully patched with a React Suspense fix
✔ Routes manifest was successfully generated
- ready started server on 0.0.0.0:3000, url: http://localhost:3000
- event compiled client and server successfully in 175 ms (20 modules)
- wait compiling...
- event compiled client and server successfully in 62 ms (20 modules)
- wait compiling / (client and server)...
- event compiled client and server successfully in 1010 ms (1317 modules)
- wait compiling...
- event compiled client and server successfully in 104 ms (1317 modules)
- wait compiling /api/list-videos (client and server)...
- event compiled successfully in 49 ms (110 modules)
- error Error [NoHostAvailableError]: All host(s) tried for query failed. First host tried, 52.59.159.247:9042: DriverError: Connection timeout
    at Socket.connectTimedOut (/home/attila/Documents/ScyllaDB/FY23Q4/daniel-video-streaming/video-streaming/node_modules/cassandra-driver/lib/connection.js:205:19)
    at Object.onceWrapper (node:events:628:28)
    at Socket.emit (node:events:514:28)
    at Socket._onTimeout (node:net:589:8)
    at listOnTimeout (node:internal/timers:573:17)
    at process.processTimers (node:internal/timers:514:7) {
  info: 'Cassandra Driver Error'
}. See innerErrors.
    at ControlConnection._borrowFirstConnection (/home/attila/Documents/ScyllaDB/FY23Q4/daniel-video-streaming/video-streaming/node_modules/cassandra-driver/lib/control-connection.js:307:15)
    at process.processTicksAndRejections (node:internal/process/task_queues:95:5)
    at async ControlConnection._initializeConnection (/home/attila/Documents/ScyllaDB/FY23Q4/daniel-video-streaming/video-streaming/node_modules/cassandra-driver/lib/control-connection.js:526:7)
    at async ControlConnection.init (/home/attila/Documents/ScyllaDB/FY23Q4/daniel-video-streaming/video-streaming/node_modules/cassandra-driver/lib/control-connection.js:212:5)
    at async Client._connect (/home/attila/Documents/ScyllaDB/FY23Q4/daniel-video-streaming/video-streaming/node_modules/cassandra-driver/lib/client.js:513:5)
    at async Client._execute (/home/attila/Documents/ScyllaDB/FY23Q4/daniel-video-streaming/video-streaming/node_modules/cassandra-driver/lib/client.js:1000:5)
    at async listVideos (webpack-internal:///(api)/./src/pages/api/list-videos.ts:10:23) {
  digest: undefined
}
- wait compiling /_error (client and server)...
- event compiled client and server successfully in 132 ms (1330 modules)
- error Error [SyntaxError]: Unexpected token '<', "<!DOCTYPE "... is not valid JSON
    at JSON.parse (<anonymous>)
    at parseJSONFromBytes (node:internal/deps/undici/undici:6662:19)
    at successSteps (node:internal/deps/undici/undici:6636:27)
    at node:internal/deps/undici/undici:1236:60
    at node:internal/process/task_queues:140:7
    at AsyncResource.runInAsyncScope (node:async_hooks:206:9)
    at AsyncResource.runMicrotask (node:internal/process/task_queues:137:8)
    at process.processTicksAndRejections (node:internal/process/task_queues:95:5) {
  digest: undefined
```

Maybe breaking up the PR into smaller, testable PRs would help

**Comment by danielhe4rt** [comment]
@zseta it's probably related to your connection with the ScyllaDB Cluster AFAIK.

Did you updated you IP into ScyllaDB Cluster allow list?

**Pull request merged**